source/h1_connection.c

```diff
@@ -721,6 +722,74 @@ static void s_client_update_incoming_stream_ptr(struct aws_h1_connection *connec
+    }
+    struct aws_http_connection *connection_base = stream->base.owning_connection;
+    struct aws_h1_connection *connection = AWS_CONTAINER_OF(connection_base, struct aws_h1_connection, base);
+    /* Time out happened, close the connection manually */
+    uint64_t idle_timeout_ms = stream->base.client_data->idle_timeout_ms == 0
+                                   ? connection_base->client_data->idle_timeout_ms
+                                   : stream->base.client_data->idle_timeout_ms;
+    AWS_LOGF_INFO(
+        AWS_LS_HTTP_CONNECTION,
+        "id=%p: Closing connection as timeout after request sent to the first byte received happened. Idle time "
+        "is %" PRIu64 " ms",
+        (void *)connection_base,
+        idle_timeout_ms);
+
+    /* Don't stop reading/writing immediately, let that happen naturally during the channel shutdown process. */
+    s_stop(
+        connection,
+        false /*stop_reading*/,
+        false /*stop_writing*/,
+        true /*schedule_shutdown*/,
+        AWS_ERROR_HTTP_REQUEST_IDLE_TIMEOUT);
+    connection_base->client_data->request_idle_timeout_task = NULL;
+
+cleanup:
+    aws_mem_release(stream->base.alloc, task);
+}
+
+static void s_request_finished_sending(struct aws_h1_stream *stream) {
+    struct aws_http_connection *connection = stream->base.owning_connection;
+    struct aws_channel *channel = aws_http_connection_get_channel(connection);
+    AWS_ASSERT(aws_channel_thread_is_callers_thread(channel));
+
+    stream->is_outgoing_message_done = true;
+    AWS_ASSERT(stream->base.metrics.send_end_timestamp_ns == -1);
+    aws_high_res_clock_get_ticks((uint64_t *)&stream->base.metrics.send_end_timestamp_ns);
+    AWS_ASSERT(stream->base.metrics.send_start_timestamp_ns != -1);
+    AWS_ASSERT(stream->base.metrics.send_end_timestamp_ns >= stream->base.metrics.send_start_timestamp_ns);
+    stream->base.metrics.sending_duration_ns =
+        stream->base.metrics.send_end_timestamp_ns - stream->base.metrics.send_start_timestamp_ns;
+    if (stream->base.metrics.receive_start_timestamp_ns == -1) {
+        /* We haven't receive any message, schedule the request idle timeout task */
+        uint64_t idle_timeout_ms = stream->base.client_data->idle_timeout_ms == 0
+                                       ? connection->client_data->idle_timeout_ms
+                                       : stream->base.client_data->idle_timeout_ms;
+        if (idle_timeout_ms != 0) {
+            connection->client_data->request_idle_timeout_task =
+                aws_mem_calloc(stream->base.alloc, 1, sizeof(struct aws_task));
+            aws_task_init(
+                connection->client_data->request_idle_timeout_task,
+                s_http_request_idle_timeout_task,
+                stream,
+                "http_request_idle_timeout_task");
+            uint64_t now_ns = 0;
+            int error = aws_channel_current_clock_time(channel, &now_ns);
+            AWS_FATAL_ASSERT(!error);
```

Comment:
I mean, it would suck if the clock failed, but an HTTP connection closing early is better than an application crashing
```suggestion
```